daemon/service.c

```diff
@@ -722,27 +790,32 @@ service_dbus_call_cb(GDBusConnection       *connection,
         }
     }
     else if( !g_strcmp0(method_name, PERMISSIONMGR_METHOD_SET_GRANTED) ) {
-        guint32      uid = SESSION_UID_UNDEFINED;
-        const gchar *app = NULL;
-        gchar **vector   = NULL;
-        g_variant_get(parameters, "(u&s^as)", &uid, &app, &vector);
-        appsettings_t *appsettings = NULL;
-        if( !control_valid_user(service_control(self), uid) ) {
-            error_reply(G_DBUS_ERROR_INVALID_ARGS, SERVICE_MESSAGE_INVALID_USER, uid);
-        }
-        else if( !(appsettings = control_appsettings(service_control(self), uid, app)) ) {
-            error_reply(G_DBUS_ERROR_INVALID_ARGS, SERVICE_MESSAGE_INVALID_APPLICATION, app);
-        }
-        else if( !vector ) {
-            error_reply(G_DBUS_ERROR_INVALID_ARGS, SERVICE_MESSAGE_INVALID_PERMISSIONS);
-        }
-        else {
-            stringset_t *granted = stringset_from_strv(vector);
-            appsettings_set_granted(appsettings, granted);
-            stringset_delete(granted);
-            value_reply(NULL);
+        if( !service_may_administrate(sender) ) {
+            error_reply(G_DBUS_ERROR_ACCESS_DENIED, SERVICE_MESSAGE_RESTRICTED_METHOD, sender,
+                        PERMISSIONMGR_METHOD_SET_GRANTED);
```

Comment:
Side note: more copy-paste accumulation... but unless all of it is refactored somehow to be cleaner, it is better to keep checks in context of actions - i.e. just as they are in this pr.

That said: it looks like copy paste -> making it copy-paste-friendly for the next guy/gal too might be beneficial to avoid not so easy to spot hiccups -> in each case we already know that 'method_name' contains what we need -> we could  use: 

    error_reply(..., sender, method_name);

Reply:
This is something I considered for a moment but didn't have a good reason to do it like this. Your reasoning is sound so I changed those.